db/view/view.cc

```diff
@@ -2623,7 +2626,137 @@ future<> view_builder::migrate_to_v2(locator::token_metadata_ptr tmptr, db::syst
 }
 
 void view_builder::upgrade_to_v2() {
+    if (_view_build_status_on_group0) {
+        return;
+    }
+
     _view_build_status_on_group0 = true;
+
+    if (_init_virtual_table_on_upgrade) {
+        init_virtual_table();
+    }
+}
+
+// policy for the multishard reader below which is used to read system.view_build_status_v2 table
+class view_build_status_policy
+    : public reader_lifecycle_policy_v2
+    , public enable_shared_from_this<view_build_status_policy> {
+
+    template <typename T>
+    using foreign_unique_ptr = foreign_ptr<std::unique_ptr<T>>;
+
+    struct reader_context {
+        foreign_ptr<lw_shared_ptr<const dht::partition_range>> range;
+        foreign_unique_ptr<utils::phased_barrier::operation> read_operation;
+        reader_concurrency_semaphore* semaphore;
+    };
+
+    distributed<replica::database>& _db;
+    table_id _table_id;
+    std::vector<reader_context> _contexts;
+public:
+    view_build_status_policy(distributed<replica::database>& db, table_id table_id)
+        : _db(db)
+        , _table_id(table_id)
+        , _contexts(smp::count) {
+    }
+    virtual mutation_reader create_reader(
+            schema_ptr schema,
+            reader_permit permit,
+            const dht::partition_range& range,
+            const query::partition_slice& slice,
+            tracing::trace_state_ptr trace_state,
+            mutation_reader::forwarding fwd_mr) override {
+        const auto shard = this_shard_id();
+        auto& cf = _db.local().find_column_family(_table_id);
+
+        _contexts[shard].range = make_foreign(make_lw_shared<const dht::partition_range>(range));
+        _contexts[shard].read_operation = make_foreign(std::make_unique<utils::phased_barrier::operation>(cf.read_in_progress()));
+        _contexts[shard].semaphore = &_db.local().get_reader_concurrency_semaphore();
+
+        return cf.as_mutation_source().make_reader_v2(
```

Comment:
This policy seems pretty standard, why was this needed (instead of using an existing one)?

Reply:
I tried few things and this is what worked for me, but I'm not very familiar with this and there's probably room for improvement.
I tried now to replace this with the multishard streaming reader but I have some issues with this, I need to look more into this

Reply:
I don't think the streaming reader will work. Instead, move the streaming reader's policy into some appropriate header, so it can be shared by this and the streaming reader. AFAIK the two policies are identical.

Reply:
done